File: backend/app/models/candle.py
from sqlalchemy import Column, String, Float, Integer, DateTime, UniqueConstraint
from app.db import Base

class Candle(Base):
    __tablename__ = "candles"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    symbol = Column(String, index=True, nullable=False)
    timeframe = Column(String, default="1D", nullable=False)
    timestamp = Column(DateTime(timezone=True), index=True, nullable=False)
    open = Column(Float, nullable=False)
    high = Column(Float, nullable=False)
    low = Column(Float, nullable=False)
    close = Column(Float, nullable=False)
    volume = Column(Float, nullable=False)
    source = Column(String, nullable=True)
    adjustment_type = Column(String, default="unadjusted")

    __table_args__ = (
        UniqueConstraint('symbol', 'timeframe', 'timestamp', 'adjustment_type', name='uq_candle'),
    )

File: frontend/src/types/candle.ts
/** OHLCV candle data from the backend */
export interface Candle {
  id: number;
  symbol: string;
  timeframe: string;
  timestamp: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
  source: string;
  adjustment_type: string;
}

/** Lightweight Charts format for candlestick rendering */
export interface ChartCandle {
  time: string;
  open: number;
  high: number;
  low: number;
  close: number;
}

/** Lightweight Charts format for volume histogram */
export interface ChartVolume {
  time: string;
  value: number;
  color: string;
}
